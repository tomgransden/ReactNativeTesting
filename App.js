/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Text,
  TextInput,
  View,
  Button,
} from 'react-native';
import {withFormik} from 'formik';

export const App = ({values, setFieldValue, errors, handleSubmit, status}) => {
  return (
    <SafeAreaView style={styles.container}>
      {status !== 'Success' ? (
        <>
          <Text style={styles.header}>Sample formik app</Text>
          <TextInput
            style={styles.usernameField}
            onChangeText={text => setFieldValue('email', text)}
            value={values.email}
            placeholder={'Username'}
            testID={'Username'}
          />
          <TextInput
            style={styles.usernameField}
            onChangeText={text => setFieldValue('password', text)}
            value={values.password}
            secureTextEntry={true}
            placeholder={'Password'}
            testID={'Password'}
          />
          {Object.keys(errors).length > 0 && (
            <View testID={'ErrorContainer'} style={styles.errorBackground}>
              <Text>{errors.email}</Text>
            </View>
          )}
          <Button
            testID={'Submit'}
            style={styles.submit}
            onPress={handleSubmit}
            title={'Submit'}
          />
        </>
      ) : (
        <Text testID={'Sucesss'}>Success</Text>
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'grey',
  },
  header: {
    fontSize: 32,
  },
  usernameField: {
    fontSize: 24,
    height: 56,
    backgroundColor: 'white',
    marginTop: 12,
  },
  errorBackground: {
    marginTop: 12,
    backgroundColor: 'red',
    height: 56,
  },
  submit: {
    marginTop: 12,
  },
});

const form = withFormik({
  displayName: 'DemoApp',
  mapPropsToValues: () => ({email: '', password: ''}),
  handleSubmit: (values, {setErrors, setStatus}) => {
    if (values.email === '' || values.password === '') {
      setErrors({email: 'Username and password are required'});
    } else {
      setStatus('Success');
    }
  },
});

export default form(App);
