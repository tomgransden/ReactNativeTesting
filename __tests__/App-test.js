/**
 * @format
 */

import 'react-native';
import React from 'react';
import {App} from '../App';

import {fireEvent, render, wait} from '@testing-library/react-native';

const baseProps = {
  handleSubmit: jest.fn(),
  setFieldValue: jest.fn(),
  values: {
    email: '',
    password: '',
  },
  errors: {},
  status: '',
};

const filledFormikProps = {
  ...baseProps,
  values: {
    email: 'MockUsername',
    password: 'MockPassword',
  },
};

const errorFormikProps = {
  ...baseProps,
  errors: {email: 'Mock email error message'},
};

const successFormikProps = {
  ...baseProps,
  values: {
    email: 'thomasg@123.com',
    password: 'Password123',
  },
  status: 'Success',
};

describe('App', () => {
  it('renders without crashing', () => {
    const {baseElement} = render(<App {...baseProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Renders form values correctly', () => {
    const {baseElement} = render(<App {...filledFormikProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Renders out errors correctly', () => {
    const {baseElement} = render(<App {...errorFormikProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Shows the success text when the form is submitted correctly', () => {
    const {baseElement} = render(<App {...successFormikProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call setFieldValue when typing username', async () => {
    const {getByTestId} = render(<App {...baseProps} />);
    const username = getByTestId('Username');
    await wait(() => fireEvent.changeText(username, 'MockUsername'));
    expect(baseProps.setFieldValue).toBeCalledWith('email', 'MockUsername');
  });

  it('Should call handleSubmit when button is pressed', async () => {
    const {getByTestId} = render(<App {...baseProps} />);
    const submit = getByTestId('Submit');

    await wait(() => fireEvent.press(submit));
    expect(baseProps.handleSubmit).toBeCalled();
  });
});
