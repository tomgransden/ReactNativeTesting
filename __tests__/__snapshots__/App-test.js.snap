// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders form values correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Object {
          "backgroundColor": "grey",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        Sample formik app
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Username"
        underlineColorAndroid="transparent"
        value="MockUsername"
      />
      <TextInput
        allowFontScaling={true}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Password"
        underlineColorAndroid="transparent"
        value="MockPassword"
      />
      <Button
        style={
          Object {
            "marginTop": 12,
          }
        }
        testID="Submit"
        title="Submit"
      />
    </SafeAreaView>
  </View>
</View>
`;

exports[`App Renders out errors correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Object {
          "backgroundColor": "grey",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        Sample formik app
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Username"
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Password"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "backgroundColor": "red",
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="ErrorContainer"
      >
        <Text>
          Mock email error message
        </Text>
      </View>
      <Button
        style={
          Object {
            "marginTop": 12,
          }
        }
        testID="Submit"
        title="Submit"
      />
    </SafeAreaView>
  </View>
</View>
`;

exports[`App Shows the success text when the form is submitted correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Object {
          "backgroundColor": "grey",
          "flex": 1,
        }
      }
    >
      <Text
        testID="Sucesss"
      >
        Success
      </Text>
    </SafeAreaView>
  </View>
</View>
`;

exports[`App renders without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Object {
          "backgroundColor": "grey",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        Sample formik app
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Username"
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 24,
            "height": 56,
            "marginTop": 12,
          }
        }
        testID="Password"
        underlineColorAndroid="transparent"
        value=""
      />
      <Button
        style={
          Object {
            "marginTop": 12,
          }
        }
        testID="Submit"
        title="Submit"
      />
    </SafeAreaView>
  </View>
</View>
`;
